### DISCUSSION TITLE
Custom connection handshake

### DISCUSSION THREAD

**Discussion by rob-maron** [libp2p/rust-libp2p#5518] [Q&A] [answered]
What's the best way to require a peer to perform additional negotiation on a new connection? For example, perhaps there is the need to authenticate by requiring the peer knows a certain secret. Another example could be exchanging application version hashes and rejecting a connection if the versions do not match. 

Here are some potential solutions I gathered:
1. Implement a `Transport` wrapper that implements `Transport` itself and requires the additional negotiation. The problem here is that, at least when wrapping QUIC, the connection is already natively multiplexed (which makes sense for QUIC). But maybe there is an easy way to open a substream on [`Connection`](https://docs.rs/libp2p-quic/0.10.2/libp2p_quic/struct.Connection.html) that I'm missing.

2. Do the same as the above, but inside of an upgrade with `InboundUpgrade` and `OutboundUpgrade`. I seem to run into the same problem as above. I had also looked at [this discussion](https://github.com/libp2p/rust-libp2p/discussions/4985#discussioncomment-7769992), which made it seem unsupported 

3. Implement a custom `NetworkBehaviour` that performs the handshake on a [`handle_established_inbound_connection`](https://docs.rs/libp2p/latest/libp2p/swarm/trait.NetworkBehaviour.html#tymethod.handle_established_inbound_connection). While I don't have direct access to the connection at this stage, I think by looking at what the `Identify` behavior does may be a good path forward.

Is there a good solution above, or is it something else entirely? Thanks for the help!

**Comment by rob-maron** [comment, marked as answer]
I ended up going with the first option, the `Transport` wrapper. With a bit of magic, you can wrap `dial` and `poll` futures to add a secondary negotiation layer:

```rust
fn dial(addr) {
    match self.inner.dial(addr) {
        Ok(original_future) => Ok(Box::pin(async move {
            // Wait the original future
            let stream = original_future.await?;

            // Open a substream for your secondary handshake
            let mut substream =
                poll_fn(|cx| stream.1.poll_outbound_unpin(cx)).await?;

            // You now have access to a stream with `AsyncRead`/`AsyncWrite`
            // .. your handshake logic ..
        }
    }
}
```

On the other side, `poll`, you would need to perform the other side of the handshake on a new incoming connection.

**Comment by jxs** [comment]
interesting! And thanks for answering your own question, btw can you describe what is your use case?
